src/sleep/sleep.c

```diff
@@ -38,8 +39,192 @@
 #include "strv.h"
 #include "util.h"
 
+#include <linux/fs.h>
+#include <linux/fiemap.h>
+
 static char* arg_verb = NULL;
 
+static struct fiemap *read_fiemap(int fd)
+{
+        struct fiemap *fiemap = NULL;
+        struct fiemap *result_fiemap = NULL;
+        struct fiemap *fm_tmp; /* need to store pointer on realloc */
+        int extents_size;
+        struct stat statinfo;
+        uint32_t result_extents = 0;
+        uint64_t fiemap_start = 0, fiemap_length;
+
+        if (fstat(fd, &statinfo) != 0) {
+                fprintf(stderr, "Cannot determine file size, errno=%d (%s)\n",
+                                errno, strerror(errno));
+                return NULL;
+        }
+        fiemap_length = statinfo.st_size;
+
+        fiemap = malloc(sizeof(struct fiemap));
+        if (fiemap == NULL) {
+                fprintf(stderr, "Out of memory allocating fiemap\n");
+                return NULL;
+        }
+
+        result_fiemap = malloc(sizeof(struct fiemap));
+        if (result_fiemap == NULL) {
+                fprintf(stderr, "Out of memory allocating fiemap\n");
+                goto fail_cleanup;
+        }
+
+        /*  XFS filesystem has incorrect implementation of fiemap ioctl and
+         *  returns extents for only one block-group at a time, so we need
+         *  to handle it manually, starting the next fiemap call from the end
+         *  of the last extent
+         */
+        while (fiemap_start < fiemap_length) {
+                memset(fiemap, 0, sizeof(struct fiemap));
+
+                fiemap->fm_start = fiemap_start;
+                fiemap->fm_length = fiemap_length;
+                fiemap->fm_flags = FIEMAP_FLAG_SYNC;
+
+                /* Find out how many extents there are */
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0) {
+                        fprintf(stderr, "fiemap ioctl() FS_IOC_FIEMAP failed, errno=%d (%s)\n",
+                                errno, strerror(errno));
+                        goto fail_cleanup;
+                }
+
+                /* Nothing to process */
+                if (fiemap->fm_mapped_extents == 0)
+                        break;
+
+                /* Result fiemap have to hold all the extents for the hole file */
+
+                /* Read in the extents */
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (fiemap->fm_mapped_extents);
+
+                /* Resize fiemap to allow us to read in the extents */
+                fm_tmp = realloc(fiemap,
+                                 sizeof(struct fiemap) + extents_size);
```

Comment:
I think it would be nice to use GREEDY_REALLOC() here, like we usually do, to exponentially grow the buffer